docs/developer-guide/partition-data.rst

```diff
@@ -10,26 +10,69 @@ of the join have the same configuration settings for partitions.
 sides of the join, are delivered to the same stream task during processing.
-If your inputs aren't co-partitioned, you need to :ref:`re-key one of the
-them <ensure-co-partitioning>` by using the PARTITION BY clause.
 
-.. _keys-have-same-keying-scheme:
+.. _keys-have-same-key-schema:
 
-Records Have the Same Keying Scheme
-===================================
+Records have the Same Key Schema
+================================
 
-The input records for the join must have the same keying scheme, which means
-that the join must use the same key field on both sides.
+For a join to work the keys from both sides must have the same serialized binary data.
 
-For example, you can join a stream of user clicks that's keyed by a ``VARCHAR userId``
-field with a table of user profiles that's keyed by a ``VARCHAR userId`` field. 
-The join won't match if the key fields don't have the same name and type.
+For example, you can join a stream of user clicks that's keyed on a ``VARCHAR`` user id with a table
+of user's profiles that's also keyed on a ``VARCHAR`` user id. Records with the exact same user id
+on both sides will be joined.
+
+Tables created on top of existing Kafka topics, for example those created with a ``CREATE TABLE``
+statement, will be keyed on the data held in the key of the records in the
+Kafka topic.  KSQL presents this data in the ``ROWKEY`` column, and expects the data to be a ``VARCHAR``.
+
+Tables created inside KSQL from other sources, for example those created with a ``CREATE TABLE AS SELECT``
+statement, will copy the key from their source(s) unless there is an explicit ``GROUP BY`` or ``PARTITION BY``
+clause, which can changed what the table is keyed on.
+
+KSQL will automatically repartition a stream of data before joining, if required. In the example below
+KSQL is used to join a stream of clicks with a table of users.  The ``clicks`` stream is not keyed on anything
+initially. The ``users`` table keyed on ``userId``.  Because the stream is being joined on a column other than
+it's key field, i.e. ``ROWKEY``, KSQL will automatically repartition the stream under the hood.
+
+.. code:: sql
+
+   -- clicks stream, with an unknown key.
+   CREATE STREAM clicks (userId STRING, url STRING) WITH(kafka_topic='clickstream', value_format='json');
+
+   -- users table, where the records have the userId in the record's key:
+   CREATE TABLE  users  (userId STRING, fullName STRING) WITH(kafka_topic='users', value_format='json', key='userId');
+
+   -- join will automatically repartition clicks stream:
+   SELECT clicks.url, users.fullName FROM clicks JOIN users ON clicks.userId = users.userId;
+
+.. note::
+
+   While KSQL will automatically repartition a stream should a join require it, KSQL will reject any join
+   any table's column that is not the key. This is because KSQL does not support joins on foreign keys
+   and repartitioning a table would corrupt the data.

```

Comment:
I would define 'corrupt'. If people don't care about order then repartitioning a table's topic is valid and talk of corruption would worry people that their data is not safe in Kafka.

Reply:
Take a look up a the updated version. Is that better?  I'm not sure if it's actually worth explaining how the ordering can cause issues?  Or how the user can manually repartition?  I'm not even sure there is a nice way to do the latter in KSQL is there?  I guess you'd have to import the topic as a stream, repartition, then reimport as a table...?  Is this documented anywhere I can refer to?